RFC-0004: Fulltext Search

fordN opened the pull request:
A proposal for a new feature to allow subgraph developers to add fulltext search indexes to their entities and thus expose fulltext search ranked filtering in the GraphQL API.

fordN committed:
Add RFC for Fulltext search indexes

fordN commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,223 @@
+    specs: String!,
+    address: Bytes!,
+    marked: Boolean!,
+    itemSearch: String! @fullTextSearchIndex(language: "english", fields: [{name: "name", weight: 5}, {name: "description", weight: 3}, {name: "specs", weight: 1}])
```
Some input on the directive name is welcome here! 
With the possibility in mind of there being additional custom indexes later,  an option I've considered is using a more general directive such as: `@index(type: String, param1: [String], param2: [String])`

yanivtal replied:
I much prefer the @index approach

Zerim replied:
I'm not a big fan of the directive approach and even less so when on top of a virtual attribute whose only purpose is to support the functioning of an index.

Indexes and data models are separate concerns, and forcing storage level concerns into the application data model hurts re-usability and limits flexibility for the Indexer.

I think I'd rather see this go somewhere in the manifest under i.e., `suggested_indexes`, where the subgraph developer could also provide hints to other indexes that they think may be useful initially and Indexers can choose to incorporate into their decisions until they have enough data on query patterns to make that decision for themselves.

fordN replied:
Given the need to apply full-text search indexes that span multiple entities, I would like to update the proposal to have the index defined as a separate type in the schema or an extension to the query type. This way the return type can be a `union` or `interface` type that includes the affected entities and a new query will be generated rather than simply adding a new filter to the existing entity query.  
Including the index definition in the manifest rather then the schema would work too, though I personally prefer the schema. See my reasoning in response to your other comment below.

yanivtal commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,223 @@
+     
+## Urgency
+
+Fulltext search filters are not essential to the value proposition of 
```
I don't see how it's true that fulltext search isn't essential to the value proposition of Graph Node. Disagree!

fordN replied:
Yep, I have no argument there; fulltext search style querying is vital to the value proposition.  

I believe what I was trying to convey is that: if the feature is not added immediately the Graph Node still has a strong value proposition and most current and prospective users will still be happy.  Maybe I'm being too literal about the difference between urgency and priority; but it's going to be a valuable feature that will likely get a lot of use, let's get it added right away! :)

yanivtal commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,223 @@
+search bars that filter data across multiple fields of an Entity.
+
+With the proposed feature I aim to:
+  - provide tools for subgraph developers to define composite search 
```
I think fulltext search indexes should be able to span across entities so this language is a little limiting.

fordN replied:
I think full text search could be very useful across entities, so I've updated this language to include that intent. 
With that goal in mind I'm working on how to enable this behavior in the GraphQL API while still encouraging data models that will give performant querying and index building.

yanivtal commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,223 @@
+    specs: String!,
+    address: Bytes!,
+    marked: Boolean!,
+    itemSearch: String! @fullTextSearchIndex(language: "english", fields: [{name: "name", weight: 5}, {name: "description", weight: 3}, {name: "specs", weight: 1}])
```
I thought we decided to move away from including this in the schema. When did we decide to put it back?

fordN replied:
When you say "move away from including this in the schema" I'm assuming you are also talking about the decision to include it in the subgraph definition in the schema? Or are you simply saying it should be in the subgraph definition, but somewhere else?

fordN replied:
I have definitely gone back and forth on putting it in the base subgraph definition or allowing it to be an add on to an already deployed subgraph, but we had a brief discussion a few weeks ago on a call; Jannis and I (if I remember correctly) agreed that it needs to be a part of the definition of uniqueness for a subgraph to keep a stable GraphQL API for each subgraph. It won't work to have a query filter or type change out from under a user if indexes are updated. With the subgraph composition though we will be able to maintain flexibility for subgraph developers to add indexes to existing subgraphs.  

Looking forward to the decentralized network, the more I thought through the added complexity of essentially adding an indexing marketplace to the query marketplace the more it seems that it poses risks to the health of the data network. 
  - Will index updates have their own system for being propagated to Graph Nodes storing that subgraph? Why not integrate them into the definition of a subgraph, so we can leverage the existing query discovery and subgraph data propagation systems.
  -  A query engine will have to find a node that not only has the subgraph deployment, but also the correct combination of indexes applied to it.  This will involve propagating index updates to the network and could result in a less clear graph of data.  
- On the other hand you could argue that forcing users to create a new subgraph in order to update or add a fulltext search index will lead to a large set of very similar subgraphs, but I _think_ the complexity of organizing that through interface design will in the end be simpler than managing different schemas/indexes on a single subgraph.  I also think the aforementioned interface work will be useful or needed for managing the complexity introduced by subgraph composition which could also leading to situations where many similar subgraphs are being created.
- Perhaps less important than correctness on the decentralized network, but keep in mind there will need to be new interfaces designed and built on Graph Explorer for managing indexes.

Concerning the choice of putting the definition in the schema: I've also thought about creating yaml files for the index definition or adding to the subgraph definition, but the schema is where the data model and query model is defined and it makes sense to keep it query concerns there.  Though it might make sense to separate raw data model concerns from front end query concerns, we are already using @derivedFrom for virtual fields that are derived from Entity data and I think keeping the query model in one place together with the data model is a useful simplification.

yanivtal replied:
Alright I see the case for this. Then can we look at these changes:
1. Switch to @index directive with a type: "fulltext"
2. Make sure the fields attributes can include fields on other entities in the schema (including imported types from subgraph composition)?

Zerim replied:
It seems like a lot of thought has gone into this and the concerns above around discovery and alignment around the right text full-text search indexes seem valid. These are definitely different from normal attribute indexes in that they are much harder to organically align on in a distributed fashion based on query access patterns (because you still need to choose what type of search index).

Even so, combining the data model for an entity w/ a specific way of searching that entity feels wrong to me... could we still include the same information in the subgraph, but somewhere else in the manifest, perhaps as a "hint" to the Indexer rather than a mandate?

Indexes provided as hints could be given a label or an ID, which end-users can then use to reference in their query to force Indexers to use a specific index in resolving a query. This concept could be extended beyond full-text search indexes as well.

fordN replied:
We could define custom indexes in the subgraph manifest (yaml) rather than in the schema, but I think that makes it more difficult to manage. 
- A new index also means a new query with specific behavior, so I'm having trouble following how it would work as just "a hint to the Indexer".  
- We currently have many indexes per subgraph; they are derived from the schema.  Why change the paradigm?
- All of our query types are currently generated from information in the schema.  Why split query definitions across files in the subgraph?
- The index definition will consist mostly of entity field names to be included these are in the schema, so working within one file (while not a huge difference) will be easier. 

Rather than give indexes a unique id (which will not be self-describing) I think we should creates a unique query for the index, so end-users can use that specific query to force the Indexer to use a specific index. Indexers can implement that index using their own custom systems, but it should always search across the fields described in the schema (or manifest based on your suggestion). The query name can be defined in the manifest by the subgraph developer.

yanivtal commented on rfcs/0004-fulltext-search.md:
````diff
@@ -0,0 +1,223 @@
+Example Usage in a query (references the example Entity definition from above section): 
+```graphql
+query {
+    exampleItems(first: 10, where: {itemSearch_rank_cd: "sarcastic"}) {
````
I dislike this API. It exposes a lot of complexity to the developer. The API should default to simple and ergonomic with more complex capabilities added as optional parameters.

fordN replied:
I've updated the API to use more general, self-describing language: `rank_cd` -> `proximity_rank`.  Keeping in mind the query can be made without any suffixes do you think this reduces the complexity you pointed out?

There is also the complexity of using operators in the filter: `&`, `|` and `<->`.  I think these operators are general and clear enough, but their usage would need to be conveyed through documentation or schema descriptions as it may not be obvious.

fordN committed:
Generalized the indexes schema directive

- The schema directive should be flexible to allow for other index types
to be applied using the same GraphQL API.  @fullTextSearch -> @index
- Usage explanation updated to reflect the more general API.
- Expose more readable, clear filter types to end user by not directly
exposing the PSQL query types: rank and rank_cd. Use self describing
names instead.

fordN committed:
Reword the goals to include the intent to allow search across entities

fordN committed:
Update GraphQL API: support polymorphic, on top level `_Schema_` type

- Update description of defining fulltext indexes in the subgraph
definition.
- Update all GraphQL example references, now including generated schema
example as well.
- Move ranking algorithm descriptions to ##Terminology.
- Include specific list of supported languages in PSQL limitations.
- Note: _Schema_ is also used for imports in composition.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,236 @@
+
+## Terminology
+
+- `lexeme`: a basic lexical unit of a language, consisting of one word or 
```
We've been using italics (`/.../`) for special terms so far. Let's use them here too, so we maintain consistence across documents.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,236 @@
+    specs: String!,
+    address: Bytes!,
+    marked: Boolean!,
+    itemSearch: String! @index(
```
If fulltext search is expected to work across entities, then this feels wrong. One idea could be to define indexes globally on the `_Schema_` type that we're introducing for e.g. schema composition:
```graphql
type _Schema_
  @index(
    index: {
      name: "search",
      language: "english",
      include: [
        { entity: "ExampleItem", fields: [ {name: "name", weight: 5 }, ... ] }
      ]
    }
  )
```

Note: In GraphQL, there is no such thing as an any type, which is why it is better not to decouple `type` and `params` and instead define possible indexes with something like
```graphql
union IndexDefinition = FulltextIndexDefinition | OtherIndexDefinition

type FulltextIndexEntityField {
  name: String!
  weight: Int!
}

type FulltextIndexEntity {
  entity: String!
  fields: [FulltextIndexEntityField!]!
}

type FulltextIndexDefinition {
  name: String!
  language: String!
  entities: [FulltextIndexEntity!]!,
}
```

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,236 @@
+        name 
+        description
+    }
+}
```
With my above suggestion, the schema-wide index name could be used as the query field. A fulltext search with the name `search` could be used as follows in queries:
```graphql
{
  ranked: search(first: 10, by: { rank: "sarcastic&tea" }, orderBy: rank, orderDirection: desc) {
    rank
    entity {
      ... on ExampleItem {
        address
        name
      }
    }
  }

  text: search(first: 10, by: { text: "foobar" }) {
    ...
    entity {
      ... on ExampleItem {
        address
        name
      }
    }
  }
}
```

One question I have is how much information the return values should carry in addition to the entity data.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,236 @@
+Fulltext search query system implementations often involve specific systems 
+for storing and querying the text documents; however, in an effort to reduce 
+system complexity and feature implementation time I propose starting with the 
+fulltext search features built in to PostgreSQL.
```
When we build query verification into the protocol, we may not be able to use Postgres for storing entity data anymore. If we build Postgres-specific fulltext search features into the protocol _now_, we may likely have to break the protocol later. @yanivtal @Zerim Thoughts?

Zerim replied:
Do we have any choice here? Are there full-text search features that are common and deterministic to all SQL databases.

For now, I'm OK w/ us having SQL or even Postgres specific full-text search functionality in the protocol w/ the understanding that there's a lot that may have to change when we eventually move to index chains and V2.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,236 @@
+work does not have to be added in order to create or update an index. 
+However, it also introduces significant added complexity: will a separate 
+query marketplace and discovery registry then be required for finding nodes 
+with your needed subgraph-index combination.  
```
I feel we should nonetheless include as little algorithmic details in what is included in the subgraph manifest (and related files). Primarily, what should be included is anything that is required to generate an API that includes fulltext search.

An approach different from declaring _indexes_ could be to declare special fulltext query fields:
```graphql
type _Schema_
  @fulltextSearchField(
    name: "search",
    entities: [ { name: "ExampleItem", fields: [ ... ] } ]
  )
```
or similar.

fordN committed:
Include ranking algorithm explanations in the ##Terminology section

Jannis commented on SUMMARY.md:
```diff
@@ -5,6 +5,7 @@
   - [Approved RFCs](./rfcs/approved.md)
     - [RFC-0001: Subgraph Composition](./rfcs/0001-subgraph-composition.md)
     - [RFC-0002: Ethereum Tracing Cache](./rfcs/0002-ethereum-tracing-cache.md)
+    - [RFC-0004: Fulltext Search](./rfcs/0002-fulltext-search.md)
```
Link and RFC number don't match.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+# RFC-0000: Template
```
This would be `RFC-0004: Fulltext Search`

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+This RFC proposes adding a fulltext search filter type to the GraphQL API 
+to allow subgraph developers to specify language-specific, lexical, 
+composite filters that end users can use in their queries. The fulltext 
+index will be created by examining all words in a document, breaking it 
```
That it's an index is an implementation detail. How about the `fulltext search feature will examine...` instead of `fulltext index will be created by examining...`?

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+## Goals & Motivation
+
+The current set of string filters available in our GraphQL API is lacking 
+the tools needed to build effective, modern interfaces. Wildcard string 
```
I don't agree with this statement. How about we say something like this instead?
> The current set of filters available in the GraphQL API is lacking fulltext search capabilities that enable searching for strings across entities and attributes efficiently. This is needed for search functionality in dApps.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+provide the tools necessary for front-end developers to implement powerful 
+search bars that filter data across multiple fields of an Entity.
+
+With the proposed feature I aim to:
```
`With the proposed feature I aim to` -> `The proposed feature aims to`

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+
+With the proposed feature I aim to:
+  - provide tools for subgraph developers to define composite search 
+    indexes that can search across multiple fields and entities, and
```
search indexes -> search APIs
or
search indexes -> search fields

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+  - integrate powerful, user friendly text search that is flexible and 
+    can be easily defined by a subgraph developer ultimately providing 
+    the tools necesssary for front end developers to create effective 
+    user interfaces without the need to middleware.
```
`the need to middleware` reads like `to middleware` is a verb. 😉 

The whole bullet point is vague and wordy for that. Perhaps we can drop it and turn the point above it into a regular paragraph.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+current deployments. However, it will represent a realization of part of 
+the long term vision for the query network. Also several high profile users 
+have communicated that it may be a conversion blocker, so implementation 
+should be prioritized. 
```
This reads as if a delay `will represent a realization of parrt of the long term vision`. Worth rephrasing.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+  relevant ones can be shown first." [- Postgres Documentation](https://www.postgresql.org/docs/11/textsearch-controls.html#TEXTSEARCH-RANKING-SEARCH-RESULTS)
+  
+  **Algorithms**:
+  - _standard ranking_: ranking based on the number of matching lexemes.     
+  - _cover density ranking_: Cover density is similar to the standard 
+    fulltext search ranking except that the proximity of matching lexemes 
+    to each other is taken into consideration. This function requires 
+    lexeme positional information to perform its calculation, so it ignores 
+    any "stripped" lexemes in the index.            
```
Can we link to specifications of these algorithms?

fordN replied:
Isn't that an implementation detail?

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+
+### Subgraph Definition
+
+Part of the power of fulltext search indexes are their flexibility, so 
```
I still don't like that we're talking about `indexes` so much here. To me, they are nothing but an implementation detail. We should be talking purely about the fulltext search API and its behavior.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+new subgraphs for the express purpose of updating fulltext search fields 
+for end-use cases. 
+
+For each fulltext search index a subgraph developer must be able to specify:
```
`index` -> API (or field)?

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+For each fulltext search index a subgraph developer must be able to specify:
+    1. a language, 
+    2. a set of text document fields to include,
+    3. relative weighting for each field in the index.
```
Drop `in the index.`?

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+    3. relative weighting for each field in the index.
+    4. a choice of ranking algorithm for sorting query result items
+
+The proposed process of adding one or more fulltext search index to an Entity 
```
`search index` -> `search API` (or field)?

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+involves adding one or more fulltext directive to the `_Schema_` type in the 
+subgraph's GraphQL schema.  Each fulltext definition will have four required
+top level parameters: `name`, `language`, `algorithm`, and `include`. 
+The fulltext search definitions will be used to generate queries on the GraphQL
```
It will not `generate queries`, it'll generate query fields in the GraphQL schema.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+The fulltext search definitions will be used to generate queries on the GraphQL
+schema that will be exposed to the end user.  
+
+Applying fulltext search indexes across entities will be a powerful 
```
`Applying fulltext search indexes` -> `Enabling fulltext search`

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+
+Applying fulltext search indexes across entities will be a powerful 
+abstraction allowing users to search across all relevant entities in one
+query. An index search across entities will by definition have polymorphic
```
`An index search across entities will...` -> `Such a search will...`

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+abstraction allowing users to search across all relevant entities in one
+query. An index search across entities will by definition have polymorphic
+results, so in that case a union type will also be generated in the schema 
+for the query result items. 
```
> results, so in that case a union type will also be generated in the schema for the query result items.

->

> results. To address this, a union type will be generated in the API schema for fulltext search results.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+results, so in that case a union type will also be generated in the schema 
+for the query result items. 
+
+Verification of the index definition will ensure that all fields referenced 
```
`Verification` is no the right term. `Validation` is. There's also frequent use of the term `index` in this paragraph.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+Verification of the index definition will ensure that all fields referenced 
+are valid String type fields. To be clear, the proposed interface will add 
+the fulltext search index definitions to the subgraph deployment hash, so 
+adding or updating a fulltext search index will require an update to the subgraph. 
```
The `To be clear` part is obvious. It's part of the schema, the schema affects the subgraph ID. We can drop the entire sentence.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+
+## Detailed Design
+
+### Subgraph Definition
```
Definition -> Schema. `Subgraph Definition` is not a term we use (other than informally perhaps).

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+  @fulltext(
+    name: "search",
+    language: "english",
+    algorithm: "ranked",
```
If we generate a `_FulltextAlgorithm` or similar enum type, we can use `ranked` instead of `"ranked"` here:
```
enum _FulltextAlgorithm {
  ranked
}
```

Jannis commented on rfcs/0004-fulltext-search.md:
````diff
@@ -0,0 +1,275 @@
+The schema generated from the above definition: 
+```graphql
+union FulltextMediaResultItem = ...
+union FulltextSearchResultItem = Band | Album | Musician
````
I think we should call these `_FulltextMediaEntity` and `_FulltextSearchEntity`.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+
+End users of the subgraph will have access to the queries generated for the
+fulltext search indexes alongside the other queries generated for each 
+entity in the subgraph.  In the case of a fulltext search index that spans
```
Queries are not generated...

fordN replied:
Do we not create items (queries) on the Query type?

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+multiple entities, [inline fragments](https://graphql.org/learn/queries/#inline-fragments) 
+may be used in the query to deal with the polymorphic result items. In the 
+front-end the `__typename` in each result item can be used to render the 
+components accordingly.  
```
Which components? We could just say `the `__typename` field can be used to distinguish the concrete entity types of the returned results`.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+front-end the `__typename` in each result item can be used to render the 
+components accordingly.  
+
+In the "text" parameter supplied to the query there will be one special 
```
`"test"` -> ``` `text` ```

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+components accordingly.  
+
+In the "text" parameter supplied to the query there will be one special 
+operator supported, the proximity operator. Several operators are available: 
```
One special operator or several operators?

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+In the "text" parameter supplied to the query there will be one special 
+operator supported, the proximity operator. Several operators are available: 
+and, or, and proximity (`&`, `|`, `<->`.) The proximity operator allows 
+one to specify max distance between search terms: 3 words apart → `<3>`. 
```
one -> clients
max distance -> the maximum distance

``` 3 words apart -> `<3>` ``` -> ``` `foo<3>bar` is equivalent to requesting that `foo` and `bar` are three words apart ideally`?

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+union FulltextSearchResultItem = Band | Album | Musician
+type Query {
+  media...
+  search(text: String!, first: Int, skip: Int, block: Int): [FulltextSearchResultItem!]!
```
`block` is not an `Int`. I think it's `Block_height`.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+Postgres ts_vector function and will be indexed using a GIN index. The subgraph 
+developer will define a ranking algorithm to be used to sort query results,
+so the end-user facing API remains easy to use without any requirement to 
+understand the ranking algorithms.   
```
The last time we discussed this, we discussed using a separate table for this data. I'm ok either way, one extra column per entity per fulltext search field could work as well. Just wanted to double-check which one we're choosing.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+        Danish, Dutch, English, Finnish, French, German, Hungarian, Italian, 
+        Norwegian, Portuguese, Romanian, Russian, Spanish, Swedish, and Turkish.
+  - select algorithms are available, and 
+  - limited to PostgreSQL storage. 
```
Do we care about Postgres here? Isn't it an implementation detail? Aren't we starting only with English initially?

fordN replied:
I agree the Postgres implementation details should be in the plan not here.  

I remember starting with English only being mentioned on the PR discussion call, but I thought it was just an option in case we need to simplify. Is there any reason not to include the option for other languages if it doesn't require more implementation work?

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+uses native Postgres fulltext features and there is a nonzero probability 
+this choice results in slower than optimal write and read times; 
+however the tradeoff in implementation time/complexity and the existence 
+of production use case testimonials tempers my apprehension here. 
```
Another risk is that the fulltext search ability will not be translatable to future phases of the network.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+however the tradeoff in implementation time/complexity and the existence 
+of production use case testimonials tempers my apprehension here. 
+
+Are there any desired "fulltext search like" capabilities that I've missed in this proposal?   
```
That would be something for the open questions section. :wink:

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+work does not have to be added in order to create or update an index. 
+However, it also introduces significant added complexity: will a separate 
+query marketplace and discovery registry then be required for finding nodes 
+with your needed subgraph-index combination.  
```
Is this a question?

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+
+An alternative design for the feature would allow more flexibility for 
+Graph Node operators in their index implementation and create a marketplace 
+for indexes. This design moved the definition of custom indexes out of the 
```
I find the mix of future and past tense here confusing. Perhaps describe the actual design here briefly. Or make it clear that it _is_ described here, if that's the case.

Jannis commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,275 @@
+the addition or update of an @index field does not require a full blockchain 
+resync, rather the index itself just needs to be rebuilt. 
+How can we allow fulltext search index (and eventually other index types) 
+updates without requiring a full subgraph resync? 
```
We can't currently. For the moment, I feel that's ok. Should we rename `@index field` to `@fulltext directive`?

Jannis commented on rfcs/approved.md:
```diff
@@ -2,3 +2,4 @@
 
 - [RFC-0001: Subgraph Composition](./0001-subgraph-composition.md)
 - [RFC-0002: Ethereum Tracing Cache](./0002-ethereum-tracing-cache.md)
+- [RFC-0004: Fulltext Search](./rfcs/0002-fulltext-search.md)
```
0002 -> 0004

fordN committed:
Update format and content in response to and inspired by PR reviews

lutter commented on rfcs/0004-fulltext-search.md:
```diff
@@ -0,0 +1,261 @@
+  )
+  @fulltext(
+    name: "search",
+    language: "english",
```
We should use [ISO 639-1](https://en.wikipedia.org/wiki/ISO_639-1) codes for the language (like `en` for English, `de` for German etc.) Everything else on the web uses them, too.

fordN replied:
I'm open to that though I think the full names are a bit more clear because they are self-describing.

lutter replied:
They are contrary to anywhere else where systems talk about human languages (e.g., HTTP, localization libraries, etc.) which all use the ISO codes. It's also not quite true that `French` is called `French` or `Spanish` `Spanish` ;)

fordN replied:
One counter example is Postgres which doesn't use the ISO codes for its fulltext search language configuration values, but I like the idea of using a standard that isn't english centric.  
I'll update the RFC to use `ISO 639-1` codes for the language value.

fordN committed:
Use ISO 639-1 codes to specify fulltext language dictionaries
